FirebaseInstallations/Source/Library/IIDMigration/FIRInstallationsIIDTokenStore.m

```diff
@@ -0,0 +1,154 @@
+
+@interface FIRInstallationsIIDTokenInfo : NSObject <NSSecureCoding>
+@property(nonatomic, nullable, copy) NSString *token;
+@end
+
+@implementation FIRInstallationsIIDTokenInfo
+
++ (BOOL)supportsSecureCoding {
+  return YES;
+}
+
+- (void)encodeWithCoder:(nonnull NSCoder *)coder {
+}
+
+- (nullable instancetype)initWithCoder:(nonnull NSCoder *)coder {
+  self = [super init];
+  if (self) {
+    _token = [coder decodeObjectOfClass:[NSString class] forKey:@"token"];
+  }
+  return self;
+}
+
+@end
+
+@interface FIRInstallationsIIDTokenStore ()
+@property(nonatomic, readonly) NSString *GCMSenderID;
+@end
+
+@implementation FIRInstallationsIIDTokenStore
+
+- (instancetype)initWithGCMSenderID:(NSString *)GCMSenderID {
+  self = [super init];
+  if (self) {
+    _GCMSenderID = GCMSenderID;
+  }
+  return self;
+}
+
+- (FBLPromise<NSString *> *)existingIIDAuthToken {
+  return [[FBLPromise onQueue:dispatch_get_global_queue(QOS_CLASS_UTILITY, 0)
+                           do:^id _Nullable {
+                             return [self IIDDefaultTokenData];
+                           }] onQueue:dispatch_get_global_queue(QOS_CLASS_UTILITY, 0)
+                                 then:^id _Nullable(NSData *_Nullable keychainData) {
+                                   return [self IIDCheckinWithData:keychainData];
+                                 }];
+}
+
+- (FBLPromise<NSString *> *)IIDCheckinWithData:(NSData *)data {
+  FBLPromise<NSString *> *resultPromise = [FBLPromise pendingPromise];
+
+  NSError *archiverError;
+  NSKeyedUnarchiver *unarchiver;
+  if (@available(iOS 11.0, *)) {
+    unarchiver = [[NSKeyedUnarchiver alloc] initForReadingFromData:data error:&archiverError];
+  } else {
+    @try {
+      unarchiver = [[NSKeyedUnarchiver alloc] initForReadingWithData:data];
+    } @catch (NSException *exception) {
+      archiverError = [FIRInstallationsErrorUtil keyedArchiverErrorWithException:exception];
+    }
+  }
+
+  if (!unarchiver) {
+    NSError *error = archiverError ?: [FIRInstallationsErrorUtil corruptedIIDTokenData];
+    [resultPromise reject:error];
+    return resultPromise;
+  }
+
+  [unarchiver setClass:[FIRInstallationsIIDTokenInfo class] forClassName:@"FIRInstanceIDTokenInfo"];
+  FIRInstallationsIIDTokenInfo *IIDTokenInfo =
+      [unarchiver decodeObjectOfClass:[FIRInstallationsIIDTokenInfo class]
+                               forKey:NSKeyedArchiveRootObjectKey];
+
+  if (IIDTokenInfo.token.length < 1) {
+    [resultPromise reject:[FIRInstallationsErrorUtil corruptedIIDTokenData]];
+    return resultPromise;
+  }
+
+  [resultPromise fulfill:IIDTokenInfo.token];
+
+  return resultPromise;
+}
+
+- (FBLPromise<NSData *> *)IIDDefaultTokenData {
+  FBLPromise<NSData *> *resultPromise = [FBLPromise pendingPromise];
+
+  NSMutableDictionary *keychainQuery = [self IIDDefaultTokenDataKeychainQuery];
+  NSError *error;
+  NSData *data = [FIRInstallationsKeychainUtils getItemWithQuery:keychainQuery error:&error];
```

Comment:
So IID put all the keychain operations in its own serial queue and would there be race conditions if FIS try to access the same keychain from another queue?

Reply:
Yeah, there might be a race, but in most of the cases there shouldn't be one. In the common case the sequence will be following:
1. IID starts default token update if needed. For that is asks FIS for FID and waits
2. FIS attempts  the migration (if needed) and potentially reads from IID keychain
3. FIS finishes all keychain operations and passes FID to IID
4. IID continues their operations. FIS does not touch the IID keychain after the migration attempt.

So the race condition may happen only when IID modifies keychain without requesting FID which it does not do in the IID with FIS version, though it may be accidentally introduced in the future.

Reply:
Please let me know if you still have concerns about it. I'll address them in a follow up PR if needed.